common/djangoapps/entitlements/management/commands/expire_old_entitlements.py

```diff
@@ -46,24 +45,22 @@ def add_arguments(self, parser):
     def handle(self, *args, **options):
         logger.info('Looking for entitlements which may be expirable.')
 
-        # This query could be optimized to return a more narrow set, but at a
-        # complexity cost. See bug LEARNER-3451 about improving it.
-        entitlements = CourseEntitlement.objects.filter(expired_at__isnull=True).order_by('id')
-
+        total = CourseEntitlement.objects.count()
         batch_size = max(1, options.get('batch_size'))
-        entitlements = Paginator(entitlements, batch_size, allow_empty_first_page=False)
```

Comment:
Would it be more effective to use the Paginator to get the left and right limits to pass into the Task?

Can you get the id's of each page and then call the task rather than doing the math yourself, which could be prone to small bugs?

Reply:
Paginator.page_range for example

https://docs.djangoproject.com/en/2.0/topics/pagination/#django.core.paginator.Paginator.page_range

Reply:
We could yeah. But Paginator is over the filtered results. So to pass a page into celery, we’d have to pass a list of all ids in the page and then have celery go look up each id. Which I didn’t know a super efficient way to do that (could just be lack of sql knowledge).

Unless you mean a Paginator over the unfiltered results, just using the class to do the math for us and not actually to pull any queries. Which, OK I wouldn’t dislike. But my first instinct was that the math was simple enough to just do ourselves.

Reply:
I see that now.  The math makes more sense in that case.